src/reducers/computeRenderedRows.js

```diff
@@ -239,3 +245,97 @@ function computeRenderedRowOffsets(state, rowRange) {
   state.rowOffsets = rowOffsetsCache;
   state.rows = bufferMapping;
 }
+
+
+/**
+ * Walk the rows to render and compute the height offsets and
+ * positions in the row buffer.
+ *
+ * NOTE (jordan) This alters state so it shouldn't be called
+ * without state having been cloned first.
+ *
+ * @param {!Object} state
+ * @param {{
+ *   endBufferIdx: number,
+ *   endViewportIdx: number,
+ *   firstBufferIdx: number,
+ *   firstViewportIdx: number,
+ * }} rowRange
+ * @private
+ */
+function computeRenderedRowOffsetsInViewport(state, rowRange) {
+  const { rowBufferSet, rowOffsetIntervalTree, storedHeights } = state;
+  const {
+    endBufferIdx,
+    endViewportIdx,
+    firstBufferIdx,
+    firstViewportIdx,
+  } = rowRange;
+
+  const renderedRowsCount = endBufferIdx - firstBufferIdx;
+  if (renderedRowsCount === 0) {
+    state.rowOffsets = {};
+    state.rows = [];
+    return;
+  }
+
+  // output for this function
+  const rows = state.rows.slice(); // state.rows
+  const rowOffsets = {}; // state.rowOffsets
+
+  // incremental way for calculating rowOffset
+  let runningOffset = rowOffsetIntervalTree.sumUntil(firstViewportIdx);
+
+  // we add the rows inside the current view port and also calculate their offsets
+  for (let rowIdx = firstViewportIdx; rowIdx < endViewportIdx; rowIdx++) {
+    // Update the offset for rendering the row
+    rowOffsets[rowIdx] = runningOffset;
+    runningOffset += storedHeights[rowIdx];
+
+    // Get position for the viewport row
+    const rowPosition = addRowToBuffer(rowIdx, rowBufferSet, firstViewportIdx, endViewportIdx, renderedRowsCount);
+    rows[rowPosition] = rowIdx;
+  }
+
+  // we also need to store the rowOffsets for rows out side the view port that still exist in the buffer
+  const rowsInBufferSet = filter(state.rows, (rowIdx) => rowBufferSet.getValuePosition(rowIdx) !== null);
+  const rowsOutsideViewPort = filter(rowsInBufferSet, (rowIdx) => !inRange(rowIdx, firstViewportIdx, endViewportIdx));
+  rowsOutsideViewPort.forEach((rowIdx) => rowOffsets[rowIdx] = state.rowOffsets[rowIdx]);
+
+  // now we modify the state with the newly calculated rows and offsets
+  state.rows = rows;
+  state.rowOffsets = rowOffsets;
+}
+
+/**
+ * Add the row to the buffer set if it doesn't exist.
+ * If addition isn't possible due to max buffer size, it'll replace an existing element outside the given range.
+ *
+ * @param {!number} rowIdx
+ * @param {!number} rowBufferSet
+ * @param {!number} startRange
+ * @param {!number} endRange
+ * @param {!number} maxBufferSize
+ *
+ * @return {?number} the position of the row after being added to the buffer set
+ * @private
+ */
+function addRowToBuffer(rowIdx, rowBufferSet, startRange, endRange, maxBufferSize) {
```

Comment:
Should computeRenderedRowOffsets also use this?

Reply:
Yep. We'll do it (I'm planning to have `addRowsToBuffer` instead) after `computeRenderedRowOffsetsInViewport` gets fixed.

Reply:
Will do this in the follow up CR, but addRowToBuffer will only be used by a single function...